### ISSUE TITLE
write_tsdb should support http as well as tcp posts

### ISSUE THREAD

**Issue by krutaw** [collectd/collectd#2022]
*   Version of collectd: 5.6.1
*   Operating system / distribution: CentOS release 6.8 (Final)

## Expected behavior

Sending information to TSDB should allow for writing not via TCP port, but also via HTTP to allow for better error correction, advanced features, etc.

## Actual behavior

Information sent fails to post if the tsdb doesn't allow TCP posts.  Error message look like thus in the logs:

[2016-10-22 02:09:54] [error] write_tsdb plugin: send failed with status -1 (Broken pipe)
[2016-10-22 02:09:54] [error] write_tsdb plugin: error with wt_send_message

## Steps to reproduce

*   step 1
Configure collectd to load the write_tsdb plugin:
LoadPlugin write_tsdb 

*   step 2
Configure collectd to send information to the tsdb:
<Plugin write_tsdb>
  <Node>
    Host           "192.168.0.98"
    Port           "4242"
    HostTags	   "tagname=tagvalue"
  </Node>
</Plugin>

*   step 3
Start collectd and tail the collectd.log (/var/log/collectd.log for me).  If the TSDB accepts TCP puts, then it will work just fine.  If the TSDB does not accept TCP puts, the error mentioned above will occur.

**Comment by krutaw** [comment]
To complete the circle, I put in a request to allow TCP put in the program in question and referenced this request in that request.

**Comment by octo** [comment]
Hi @krutaw,

thanks for your feature request. Supporting HTTP will mean pulling in *libcurl* with non-trivial code additions and an additional dependency. Could you please elaborate on the benefits this will have for the users of this plugin? "better error correction" sounds vague and unlikely.

Best regards,
—octo

**Comment by krutaw** [comment]
@octo Absolutely.  Check out the OpenTSDB documentation: http://opentsdb.net/docs/build/html/user_guide/writing.html.  Excerpt: "The Telnet method of writing is discouraged as it doesn't provide a way of determining which data points failed to write due to formatting or storage errors. Instead use the HTTP API."  In short, writing to the telnet port (tcp) is a brute force way of sending the data but doesn't allow for proper error correction on a per data point perspective.

As far as the benefits, there are quite a few things that collectd does very, very well and with that being the case there's no reason why it can't continue to be used with other systems.   There are, however, multiple examples of other systems that allow sending of TSDB data that don't allow TCP puts due to the lack of error correction on a per data point basis.

**Comment by kev009** [comment]
I think the question to answer first is, what do you do with enhanced failure notification?  If there is nothing done to enhance the error reporting etc it is not worth the complexity IMHO, although there is a small argument for punching through firewalls easier.  I would suggest using a richer language for HTTP since you are already trading performance for the more verbose protocol.  https://github.com/wtelecom/collectd-opentsdb is one example.
